Introducing WithAttributesLazy

boekkooi-impossiblecloud opened the pull request:
Good day,

While checking our pprof profiles with a noop tracer we notice that we still allocate attributes for each `tracer.Start` call including even when these attributes will never be used.

This PR propose to introduce a `WithAttributesLazy` option. This option accepts a function that returns the required attributes.
This means that when using a [noop tracer](https://github.com/open-telemetry/opentelemetry-go/blob/86df763d90b8aee4154196ec72a9f68b4121f9c7/trace/noop/noop.go#L54) or when using the [auto tracer](https://github.com/open-telemetry/opentelemetry-go/blob/b7ca1e18c4eac3471b27d925519eb7c4ff4f58d7/trace/auto.go#L59) `Start `when `sampling == false` the attributes function will not be called as `NewSpanStartConfig` is never called reducing allocation an time spend.

I'm not sure if this is the right way to propose this change but I thought I would give it a try.
Please let me know what you think and thank you for reviewing this PR!

--- 

In regards to performance using `WithAttributesLazy` results in 1 less allocation when `NeverSample` is used. 
However, in production environment the impact can be bigger as strings don't need to be allocated and other logic can be avoided when never sampling.

```
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
cpu: AMD Ryzen 7 Pro 7735U with Radeon Graphics     
BenchmarkStartEndSpanWithAttributes
BenchmarkStartEndSpanWithAttributes/AlwaysSample
BenchmarkStartEndSpanWithAttributes/AlwaysSample-16      	 1280181	       924.1 ns/op	    1528 B/op	       7 allocs/op
BenchmarkStartEndSpanWithAttributes/NeverSample
BenchmarkStartEndSpanWithAttributes/NeverSample-16       	 2338800	       505.9 ns/op	     824 B/op	       6 allocs/op
BenchmarkStartEndSpanWithAttributesLazy
BenchmarkStartEndSpanWithAttributesLazy/AlwaysSample
BenchmarkStartEndSpanWithAttributesLazy/AlwaysSample-16  	 1314256	       903.4 ns/op	    1504 B/op	       6 allocs/op
BenchmarkStartEndSpanWithAttributesLazy/NeverSample
BenchmarkStartEndSpanWithAttributesLazy/NeverSample-16   	 2424777	       493.9 ns/op	     800 B/op	       5 allocs/op
```

pellared commented:
@boekkooi-impossiblecloud, thanks for the proposal.

I think we can look at it and discuss during next OTel Go SIG meeting (feel free to join).

Before that I think you can add some benchmarks to the PR description. I would also rename the option to `WithAttributesLazy`. Would we want also to add the same option for `go.opentelemetry.io/otel/metric` and `go.opentelemetry.io/otel/log`?

boekkooi-impossiblecloud commented:
Hey @pellared 

Thanks for the review, I added a benchmark in https://github.com/open-telemetry/opentelemetry-go/pull/6649/commits/1f48a1a77bf9f631da2f856d5773eb2c091433d3 and put the results in the PR description, I have also renamed the method to `WithAttributesLazy`.

In regards to adding `WithAttributesLazy` to metrics, I expect that most people will run with metrics on and as there is no sampling I think adding it has little value. For logging I don't know as I have no experience with it.

pellared commented:
@boekkooi-impossiblecloud, I just want to mention that I was not able to join the last few SIG meetings and we are also skipping this week. Hopefully we will discuss it on 2025-05-15. @open-telemetry/go-approvers, do you have any feedback beforehand?

boekkooi-impossiblecloud commented:
Hey @pellared 

So adding  [Enabled](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#enabled) to [Tracer][https://pkg.go.dev/go.opentelemetry.io/otel/trace#Tracer] would allow one to not create a Span at all.
However, this will not take sampling (like in the [autoTracer]https://github.com/open-telemetry/opentelemetry-go/blob/b7ca1e18c4eac3471b27d925519eb7c4ff4f58d7/trace/auto.go#L72)) into account. 

In my opinion the `WithAttributesLazy` concept works for more use-cases then the `Enabled` concept as tracers (like the [autoTracer]https://github.com/open-telemetry/opentelemetry-go/blob/b7ca1e18c4eac3471b27d925519eb7c4ff4f58d7/trace/auto.go#L72)) where sampling is used can collected the attributes when a span is sampled and other wise ignore it.

Another option that could work is doing something similar to the [zap](https://pkg.go.dev/go.uber.org/zap) [`Logger.Check`](https://pkg.go.dev/go.uber.org/zap#Logger.Check) where a type of span builder is returned which can then be started.
However, I expect that this would require a lot more coordination to get consensus on and would be a lot harder to adopt at the moment.

Let me know what you think :smile: 

P.S. I will try to join the meeting on 2025-05-15.

pellared commented:
> However, this will not take sampling (like in the [autoTracer](https://github.com/open-telemetry/opentelemetry-go/blob/b7ca1e18c4eac3471b27d925519eb7c4ff4f58d7/trace/auto.go#L72)) into account.

@boekkooi-impossiblecloud, have you considered adding attributes using the following pattern?

```go
if span.SpanContext().IsSampled() {
	span.SetAttributes(...)
}
```

pellared commented:
I created https://github.com/open-telemetry/opentelemetry-specification/issues/4501

boekkooi-impossiblecloud commented:
@pellared You are totally right using `IsRecording` combined with `SetAttributes` would also work.

Maybe it's better to close this PR and add some thing to the docs to mention this method?
```
if span.IsRecording() {
	span.SetAttributes(...)
}
```

pellared commented:
> Maybe it's better to close this PR and add some thing to the docs to mention this method?

I support this

boekkooi-impossiblecloud commented:
@pellared do you have any suggestion on where in the docs to best mention this?

pellared commented:
> @pellared do you have any suggestion on where in the docs to best mention this?

Maybe a new separate paragraph in package docs above https://pkg.go.dev/go.opentelemetry.io/otel/trace#hdr-API_Implementations? Here is something for inspiration: https://pkg.go.dev/log/slog#hdr-Performance_considerations
2.
